Secure mqtt

TheHackmeister opened the pull request:
This secures MQTT communication between NodeRED, Home Assistant, and Rabbit MQ. There will be more changes needed to communicate with ESPs, but that'll go in a different branch.

greatbacon commented on nr/data/settings.js:
```diff
@@ -227,7 +227,7 @@ module.exports = {
             // trace - record very detailed logging + debug + info + warn + error + fatal errors
             // off - turn off all logging (doesn't affect metrics or audit)
-            level: "info",
+            level: "trace",
```
I believe convention is to leave the master (i.e. production) settings for things like logs at `info` level. Configuring environments for stuff like this might be a good feature down the road too.

greatbacon commented on nr/data/.flows.json.backup:
```diff
@@ -0,0 +1 @@
+[{"id":"ee1936cd.b6d21","type":"tab","label":"Flow 1","disabled":false,"info":""},{"id":"85d0c327.d3e95","type":"mqtt-broker","z":"","name":"MQTT","broker":"mqtt","port":"8883","tls":"62b8b95c.26b7d8","clientid":"nr.scifi.farm","usetls":true,"compatmode":false,"keepalive":"60","cleansession":true,"willTopic":"","willQos":"0","willPayload":"","birthTopic":"","birthQos":"0","birthPayload":""},{"id" ...
```
Is this like, default test data for making sure the MQ stuff works?

TheHackmeister replied:
Yeah, basically. It's not really test data though, it's just a flow with a MQTT in and out node. If it connects, it'll show "connected" next to the node in NodeRED. More than that, this is just a backup of .flows.json, which is actually what's getting used.

I don't think it's necessary, so I've removed this file and added *.backup to .gitignore.

greatbacon commented on install.sh:
```diff
@@ -88,20 +85,42 @@ else
     vault_i operator unseal $unsealKey
     vault_i login $rootToken
+
+    # Configure root CA
```
I'm sure that as this project grows, the install script will also grow with it. It might be a good idea to break the script out into separate methods or even better, separate scripts if you can.

TheHackmeister commented:
Thanks for the feedback! You are 100% right about the installer needing some refactoring. I went to add some more to it, and had trouble figuring out where my additions needed to go. 

What I've done:
- Split out the script into multiple files. install.sh loads all the files in installer/bash and then runs the main installation tasks. The tasks and general functions are broken out by what component is being accessed. installer/bash/vault.sh for example, has all the vault related code and installer/bash/host.sh contains all the modifications needed to the host machine, as well as commands to manipulate Docker. 
    1. Right now, the general functions and actual install steps for a given component are all in one file, and I'm wondering if it makes sense to split them out further so that the actual install steps in a separate file. 
- Created a couple functions (extract_from_json and create_volume) for tasks that were repeated. 
- Transitioned to using a username and password generated by Vault's Rabbit MQ secrets backend for MQTT authentication. This is because I discovered that my plan of using TLS certs for MQTT authentication wasn't supported on the ESP8266.
- Configured Home Assistant to use those usernames and passwords. 
- Added volume creation for ha_db and mqtt to allow for persistence as well as to avoid having to mount things from the host file system. 
- Installer undoes previous installer's work. This allows me to run "sudo -E ./install.sh" and it will automatically bring down the old stack, and remove the old volumes. This helped speed up development a lot. 

What's left:

- [ ] Change NodeRED's MQTT authentication to use username/password.
- [ ] Create flows to renew/regenerate MQTT usernames and passwords before they expire.
- [ ] A fair number of # TODO:'s scattered throughout the code. They are mostly moving from static wait times to continuing once some condition has been detected. 
- [ ] Write up documentation.

TheHackmeister commented:
This wraps up most of the changes necessary for the infrastructure to communicate securely. MQTT is still open on port 1883, which is insecure. Once the ESP8266 is able to use TLS, that will be closed. Documentation will be added in a different PR.
